verifiers/internal/vsa/verifier.go

```diff
@@ -0,0 +1,215 @@
+	// 		"456": true,
+	// 	}
+	// }
+	allVSASubjectDigests := make(map[string]map[string]bool)
+	for _, subject := range vsa.Subject {
+		for digestType, digestValue := range subject.Digest {
+			if _, ok := allVSASubjectDigests[digestType]; !ok {
+				allVSASubjectDigests[digestType] = make(map[string]bool)
+			}
+			allVSASubjectDigests[digestType][digestValue] = true
+		}
+	}
+	if len(allVSASubjectDigests) == 0 {
+		return fmt.Errorf("%w: no subject digests found in the VSA", serrors.ErrorInvalidDssePayload)
+	}
+	// search for the expected digests in the VSA
+	for _, expectedDigest := range *vsaOpts.ExpectedDigests {
+		parts := strings.SplitN(expectedDigest, ":", 2)
+		if len(parts) != 2 {
+			return fmt.Errorf("%w: expected digest %s is not in the format <digest type>:<digest value>", serrors.ErrorInvalidDssePayload, expectedDigest)
+		}
+		digestType := parts[0]
+		digestValue := parts[1]
+		if _, ok := allVSASubjectDigests[digestType]; !ok {
+			return fmt.Errorf("%w: expected digest not found: %s", serrors.ErrorMissingSubjectDigest, expectedDigest)
+		}
+		if _, ok := allVSASubjectDigests[digestType][digestValue]; !ok {
+			return fmt.Errorf("%w: expected digest not found: %s", serrors.ErrorMissingSubjectDigest, expectedDigest)
+		}
+	}
+	return nil
+}
+
+// matchVerifierID checks if the verifier ID in the VSA matches the expected value.
+func matchVerifierID(vsa *vsa10.VSA, vsaOpts *options.VSAOpts) error {
+	if vsa.Predicate.Verifier.ID == "" {
+		return fmt.Errorf("%w: no verifierID found in the VSA", serrors.ErrorEmptyRequiredField)
+	}
+	if *vsaOpts.ExpectedVerifierID != vsa.Predicate.Verifier.ID {
+		return fmt.Errorf("%w: verifier ID mismatch: wanted %s, got %s", serrors.ErrorMismatchVerifierID, *vsaOpts.ExpectedVerifierID, vsa.Predicate.Verifier.ID)
+	}
+	return nil
+}
+
+// matchResourceURI checks if the resource URI in the VSA matches the expected value.
+func matchResourceURI(vsa *vsa10.VSA, vsaOpts *options.VSAOpts) error {
+	if vsa.Predicate.ResourceURI == "" {
+		return fmt.Errorf("%w: no resourceURI provided", serrors.ErrorEmptyRequiredField)
+	}
+	if *vsaOpts.ExpectedResourceURI != vsa.Predicate.ResourceURI {
+		return fmt.Errorf("%w: resource URI mismatch: wanted %s, got %s", serrors.ErrorMismatchResourceURI, *vsaOpts.ExpectedResourceURI, vsa.Predicate.ResourceURI)
+	}
+	return nil
+}
+
+// confirmVerificationResult checks that the policy verification result is "PASSED".
+func confirmVerificationResult(vsa *vsa10.VSA) error {
+	if normalizeString(vsa.Predicate.VerificationResult) != "PASSED" {
+		return fmt.Errorf("%w: verification result is not Passed: %s", serrors.ErrorInvalidVerificationResult, vsa.Predicate.VerificationResult)
+	}
+	return nil
+}
+
+// matchVerifiedLevels checks if the verified levels in the VSA match the expected values.
+func matchVerifiedLevels(vsa *vsa10.VSA, vsaOpts *options.VSAOpts) error {
+	vsaLevels := make(map[string]bool)
+	for _, level := range vsa.Predicate.VerifiedLevels {
```

Comment:
nit: iirc, only one value per track ios allowed in the VSA. You may want to validate that's the case. (+ unit tests)

Reply:
The specs says

> Users MUST NOT include more than one level per SLSA track.

I was originally going to leave out this inference because it didn't seem necessary, similar to how it also says

> The result of evaluating an artifact (or set of artifacts) against SLSA. SHOULD be one of these values:
> SLSA_BUILD_LEVEL_UNEVALUATED
> SLSA_BUILD_LEVEL_0
> ...

I see now that MUST and SHOULD are different, according to [how the spec use those words](https://www.rfc-editor.org/rfc/rfc2119). 

I change my mind now, I'll add it in.

Reply:
added it.